src/Storages/IStorageCluster.cpp

```diff
@@ -104,15 +104,19 @@ void ReadFromCluster::applyFilters(ActionDAGNodes added_filter_nodes)
     if (filter_actions_dag)
         predicate = filter_actions_dag->getOutputs().at(0);
 
-    createExtension(predicate);
+    auto max_replicas_to_use = static_cast<UInt64>(cluster->getShardsInfo().size());
+    if (context->getSettingsRef()[Setting::max_parallel_replicas] > 1)
+        max_replicas_to_use = std::min(max_replicas_to_use, context->getSettingsRef()[Setting::max_parallel_replicas].value);
+
+    createExtension(predicate, max_replicas_to_use);
```

Comment:
One thing I wasn't quite sure is how the the initialization works here in the `applyFilters` - not sure what is the code path to get here.

I copied parts from below to get the same number of replicas, so hopefully that will work ok.